Lib/test/test_capi/test_set.py

```diff
@@ -213,3 +216,48 @@ def test_clear(self):
             clear(object())
         self.assertImmutable(clear)
         # CRASHES: clear(NULL)
+
+
+class TestInternalCAPI(BaseSetTests, unittest.TestCase):
+    def test_set_update(self):
+        update = _testinternalcapi.set_update
+        for cls in (set, set_subclass):
+            for it in ('ab', ('a', 'b'), ['a', 'b'],
+                       set('ab'), set_subclass('ab'),
+                       frozenset('ab'), frozenset_subclass('ab')):
+                with self.subTest(cls=cls, it=it):
+                    instance = cls()
+                    self.assertEqual(update(instance, it), 0)
+                    self.assertEqual(instance, {'a', 'b'})
+                    instance = cls(it)
+                    self.assertEqual(update(instance, it), 0)
+                    self.assertEqual(instance, {'a', 'b'})
+            with self.assertRaisesRegex(TypeError, 'object is not iterable'):
+                update(cls(), 1)
+            with self.assertRaisesRegex(TypeError, "unhashable type: 'dict'"):
+                update(cls(), [{}])
+        with self.assertRaises(SystemError):
+            update(object(), 'ab')
+        self.assertImmutable(update, 'ab')
+        # CRASHES: update(NULL, object())
+        # CRASHES: update(instance, NULL)
+        # CRASHES: update(NULL, NULL)
+
+    def test_set_next_entry(self):
+        set_next = _testinternalcapi.set_next_entry
+        for cls in (set, set_subclass, frozenset, frozenset_subclass):
+            with self.subTest(cls=cls):
+                instance = cls([1, 2, 3])
+                pos = 0
+                while True:
+                    res = set_next(instance, pos)
+                    if res is None:
+                        break
+                    rc, pos, hash_, item = res
+                    self.assertEqual(rc, 1)
+                    self.assertIsInstance(pos, int)
+                    self.assertIn(item, instance)
```

Comment:
Also store items, and compare the accumulated result with `list(instance)`. It will test that

1. All items are yielded.
2. Every item was yielded only once.
3. The order of iteration is the same as in normal iteration (and not reversed order or insertion order, for example).